miden-crypto/src/merkle/smt/large_forest/history/mod.rs

```diff
@@ -0,0 +1,430 @@
+    #[must_use]
+    pub fn node_value(&self, index: &NodeIndex) -> Option<&Word> {
+        self.history
+            .deltas
+            .iter()
+            .skip(self.version_ix)
+            .find_map(|v| v.nodes.get(index))
+    }
+
+    /// Gets the value of the entire leaf in the history at the specified `index`, or returns `None`
+    /// if the version does not overlay the current tree at that leaf.
+    ///
+    /// # Complexity
+    ///
+    /// The computational complexity of this method is linear in the number of versions due to the
+    /// need to traverse to find the correct overlay value.
+    #[must_use]
+    pub fn leaf_value(&self, index: &LeafIndex<SMT_DEPTH>) -> Option<&CompactLeaf> {
+        self.history
+            .deltas
+            .iter()
+            .skip(self.version_ix)
+            .find_map(|v| v.leaves.get(index))
+    }
+
+    /// Queries the value of a specific key in a leaf in the overlay, returning:
+    ///
+    /// - `None` if the version does not overlay that leaf in the current tree,
+    /// - `Some(None)` if the version does overlay that leaf but the compact leaf does not contain
+    ///   that value,
+    /// - and `Some(Some(v))` if the version does overlay the leaf and the key exists in that leaf.
+    ///
+    /// # Complexity
+    ///
+    /// The computational complexity of this method is linear in the number of versions due to the
+    /// need to traverse to find the correct overlay value.
+    #[must_use]
+    pub fn value(&self, key: &Word) -> Option<Option<&Word>> {
+        self.leaf_value(&LeafIndex::from(*key)).map(|leaf| leaf.get(key))
+    }
+}
+
+// DELTA
+// ================================================================================================
+
+/// A delta for a state `n` represents the changes (to both nodes and leaves) that need to be
+/// applied on top of the state `n + 1` to yield the correct tree for state `n`.
+///
+/// # Cumulative Deltas and Temporal Ordering
+///
+/// In order to best represent the history of a merkle tree, these deltas are constructed to take
+/// advantage of two main properties:
+///
+/// - They are _cumulative_, which reduces their practical memory usage. This does, however, mean
+///   that querying the state of older blocks is more expensive than querying newer ones.
+/// - Deltas are applied in **temporally reversed order** from what one might expect. Most
+///   conventional applications of deltas bring something from the past into the future through
+///   application. In our case, the application of one or more deltas moves the tree into a **past
+///   state**.
+///
+/// # Construction
+///
+/// While the [`Delta`] type is visible in the interface of the history, it is only intended to be
+/// constructed by the history. Users should not be allowed to construct it directly.
+#[derive(Clone, Debug, PartialEq)]
+struct Delta {
+    /// The root of the tree in the `version` corresponding to this delta.
```

Comment:
this means post applying the changes, this is the root we get? How do we ensure the delta is applied to the right "baseline"? Would it be good to have a second `Word`/digest to ensure it's targeting the correct tree state (read: runtime footgun check)?

Reply:
Clarified the docs, but I want to avoid too many checks in the fast path for now. If this turns out to be a problem it can be added.